base_export_async/models/delay_export.py

```diff
@@ -0,0 +1,120 @@
+            raise Warning(_("The user doesn't have an email address."))
+
+        model = self.env[model_name].with_context(
+            import_compat=import_compat, **context)
+        records = model.browse(ids) or model.search(
+            domain, offset=0, limit=False, order=False)
+
+        if not model._is_an_ordinary_table():
+            fields_name = [field for field in fields_name
+                           if field['name'] != 'id']
+
+        field_names = [f['name'] for f in fields_name]
+        import_data = records.export_data(
+            field_names, raw_data).get('datas', [])
+
+        if import_compat:
+            columns_headers = field_names
+        else:
+            columns_headers = [val['label'].strip() for val in fields_name]
+
+        if export_format == 'csv':
+            csv = CSVExport()
+            result = csv.from_data(columns_headers, import_data)
+        else:
+            xls = ExcelExport()
+            result = xls.from_data(columns_headers, import_data)
+
+        export_record = self.sudo().create({'user_id': user.id})
+
+        name = "{}.{}".format(model_name, export_format)
+        attachment = self.env['ir.attachment'].create({
+            'name': name,
+            'datas': base64.b64encode(result),
+            'datas_fname': name,
+            'type': 'binary',
+            'res_model': self._name,
+            'res_id': export_record.id,
+        })
+
+        url = "{}/web/content/ir.attachment/{}/datas/{}?download=true".format(
+            self.env['ir.config_parameter'].sudo().get_param('web.base.url'),
+            attachment.id,
+            attachment.name,
+        )
+
+        time_to_live = self.sudo().env. \
+            ref('base_export_async.attachment_time_to_live').value
+        date_today = fields.Date.from_string(fields.Date.today())
+        expiration_date = fields.Date.to_string(
+            date_today + relativedelta(days=+int(time_to_live)))
+
+        odoo_bot = self.env.ref("base.partner_root")
+        email_from = odoo_bot.email
+        self.env['mail.mail'].create({
+            'email_from': email_from,
+            'reply_to': email_from,
+            'email_to': user.email,
+            'subject': _("Export {} {}").format(
+                model_name, fields.Date.to_string(fields.Date.today())),
+            'body_html': _("""
+                <p>Your export is available <a href="{}">here</a>.</p>
+                <p>It will be automatically deleted the {}.</p>
+                <p>&nbsp;</p>
+                <p><span style="color: #808080;">
+                This is an automated message please do not reply.
+                </span></p>
+                """).format(url, expiration_date),
+            'auto_delete': True,
+        })
+
+    @api.model
+    def cron_delete(self):
+        time_to_live = self.env. \
+            ref('base_export_async.attachment_time_to_live').value
+        date_today = fields.Date.from_string(fields.Date.today())
```

Comment:
with odoo 12, `fields.Date.today()` return a `datetime.date` object. the conversion is not needed here.